content/6_Advanced/Wavelet.mdx

```diff
@@ -32,9 +30,232 @@ Like a segment tree on values rather than indices.
 	</Resource>
 </Resources>
 
-### Solution - Range K-th Smallest
+Suppose you want to support the following queries:
+
+- Given a range $l$, $r$ count the number of occurrences of value $x$.
+- Given a range $l$, $r$ find the $k$ smallest element
+
+With wavelet tree, you can easily support those queries in $log(M)$ time.
+when M is the maximum value in the array.
+
+## Wavelet tree structure
+
+A wavelet tree is a binary tree where each node represents a range of values.
+The root node covers the entire range, and each subsequent level splits the
+range into two halves.
+
+We are going to maintain a segment tree over values instead of indices. Each segment will contain indices whose values lie within the segment's range. We'll save those indices in a vector. Notice that an index can be in at most $log(M)$ segments
+
+<Spoiler title="Wavelet tree Visualization">
+
+Let's say our array is: $[3,5,3,1,2,2,3,4,5,5]$
+Each node has an array representing the indecies of every number between l and r
+
+![Wavelet Tree Visualization](./assets/diagram.png)
+</Spoiler>
+
+## Solving the first type of query
+**Given a range l, r count the number of occurrences of value x.**
+
+To calculate the number of occurrences from $𝑙$ to $𝑟$, we can use the following
+formula:
+
+$$
+\begin{aligned}
+\texttt{occurrences}(l, r) = \texttt{occurrences}(r) - \texttt{occurrences}(l)
+\end{aligned}
+$$
+
+This reduces the problem to counting the number of occurrences in a prefix.
+
+One way to solve the problem is to go to the leaf node
+and perform a binary search for the number of indices less than $𝑟$
+However, let's explore a different approach that can also be extended to the
+second type of query.
+
+**A different way to find the Index of $𝑟$ in the list of vertices**
+
+Instead of binary searching on the leaf, we update $𝑟$ as we recurse down the
+tree.
+If we can determine the position (index) of $𝑟$ in the left and right
+children of a node.
+We can recurse down the tree and determine its position in the leaf node.
+
+To find the position of $𝑟$ in a node's left and right children, we need to
+determine how many indices are smaller than the middle value (mid) and precede
+$𝑟$.
+This can be done using a prefix sum.
+
+Let's define:
+- $c[i]$ = as 1 if $index[i]$ is smaller than mid otherwise 0
+- $prefixB[i]$ as prefix sum of $c[i]$
+
+Formally
+
+$$
+c[i] = index[i] < mid ? 1 : 0;
+prefixB[i] = prefixB[i - 1] + c[i]
+$$
```

Comment:
i would suggest getting rid of the formally part if you're just rewriting your statement in code

Reply:
I thought it might help people who don't fully understand what's written, or make it a bit easier to comprehend.

Reply:
```suggestion
Let's define:
- $c[i]$ = as $1$ if $index[i]$ is smaller than mid otherwise $0$
- $prefixB[i]$ as prefix sum of $c[i]$

Formally

$$
c[i] = index[i] < mid ? 1 : 0;
prefixB[i] = prefixB[i - 1] + c[i]
$$
```

Hm, to me it feels a little redundant. If you want to keep it, it would be better in a code block instead of LaTeX.